File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    private ListNode reverseLinkedList(ListNode head, int k){

        ListNode curr = head, prev = null, next;


        while(k > 0 && curr != null){
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
            k--;
        }

        return prev;
    }
    public ListNode reverseKGroup(ListNode head, int k) { // 1 2 3 4 5
        ListNode curr = head; // h - 1 2 3 4 5
        int count = 0;

        while(count < k && curr != null){ // 0 1
            curr = curr.next;
            count++;
        }

        if(count == k){
            ListNode reversedHead = reverseLinkedList(head, k); // 2 1 3 4 5

            head.next = this.reverseKGroup(curr, k); // 3 4 5
            return reversedHead;
        }

        return head;
        
    }
}

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseKGroup(ListNode head, int k) {
        ListNode curr = head;
        int currCount = 0;
        ListNode currHead = curr;
        ListNode dummy = new ListNode(0, head);
        ListNode prevGroupTail = dummy;

        while(curr != null) {
            currCount++;
            if(currCount == k) {
                ListNode saveNext = curr.next;
                curr.next = null;
                ListNode reversed = reverseListNode(currHead);
                prevGroupTail.next = reversed;
                currHead.next = saveNext;
                prevGroupTail = currHead;
                currHead = saveNext;
                currCount = 0;
                curr = saveNext;
            }
            else {
                curr = curr.next;
            }
        }
        return dummy.next;

    }

    private ListNode reverseListNode(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;

        while(curr != null) {
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }
}